src/rdf4cpp/rdf/IRIFactory.cpp

```diff
@@ -0,0 +1,360 @@
+
+#include <rdf4cpp/rdf/util/CharMatcher.hpp>
+
+namespace rdf4cpp::rdf {
+
+IRIView::IRIView(std::string_view iri) noexcept : data(iri) {
+}
+
+std::string_view IRIView::apply(const IRIView::IRIPart &r) const noexcept {
+    return data.substr(r.start, r.len);
+}
+std::optional<std::string_view> IRIView::apply_opt(const IRIView::IRIPart &r) const noexcept {
+    if (r.defined)
+        return apply(r);
+    return std::nullopt;
+}
+
+IRIView::IRIPart IRIView::get_scheme_part() const noexcept {
+    auto c = data.find(':');
+    if (c == std::string_view::npos)
+        return {0, 0, false};
+    auto s = data.find('/');
+    if (c > s)
+        return {0, 0, false};
+    return {0, c, true};
+}
+std::optional<std::string_view> IRIView::scheme() const noexcept {
+    return apply_opt(get_scheme_part());
+}
+bool IRIView::is_relative() const noexcept {
+    return !get_scheme_part().defined;
+}
+
+IRIView::IRIPart IRIView::get_authority_part(const IRIPart &scheme) const noexcept {
+    auto d = data;
+    auto s = scheme.start + scheme.len;
+    if (s > 0) {
+        ++s;
+        d = d.substr(s);
+    }
+    if (!d.starts_with("//"))
+        return {s, 0, false};
+    d = d.substr(2);
+    s += 2;
+    auto e = d.find_first_of("/#?");
+    if (e == std::string_view::npos)
+        e = d.length();
+    return {s, e, true};
+}
+std::optional<std::string_view> IRIView::authority() const noexcept {
+    return apply_opt(get_authority_part(get_scheme_part()));
+}
+
+IRIView::IRIPart IRIView::get_path_part(const IRIPart &auth) const noexcept {
+    auto b = auth.start + auth.len;
+    auto d = data.substr(b);
+    if (auth.len > 0 && !d.starts_with('/'))
+        return {b, 0, true};
+    auto e = d.find_first_of("#?");
+    if (e == std::string_view::npos)
+        e = d.length();
+    return {b, e, true};
+}
+std::string_view IRIView::path() const noexcept {
+    return apply(get_path_part(get_authority_part(get_scheme_part())));
+}
+
+IRIView::IRIPart IRIView::get_query_part(const IRIPart &path) const noexcept {
+    auto b = path.start + path.len;
+    auto d = data.substr(b);
+    if (!d.starts_with('?'))
+        return {b, 0, false};
+    d = d.substr(1);
+    ++b;
+    auto e = d.find_first_of('#');
+    if (e == std::string_view::npos)
+        e = d.length();
+    return {b, e, true};
+}
+std::optional<std::string_view> IRIView::query() const noexcept {
+    return apply_opt(get_query_part(get_path_part(get_authority_part(get_scheme_part()))));
+}
+
+IRIView::IRIPart IRIView::get_fragment_part(const IRIPart &query) const noexcept {
+    auto b = query.start + query.len;
+    auto d = data.substr(b);
+    if (!d.starts_with('#'))
+        return {b, d.length(), false};
+    return {b + 1, d.length() - 1, true};
+}
+std::optional<std::string_view> IRIView::fragment() const noexcept {
+    return apply_opt(get_fragment_part(get_query_part(get_path_part(get_authority_part(get_scheme_part())))));
+}
+
+std::tuple<std::optional<std::string_view>,
```

Comment:
I think a dedicated struct would be more readable. They should be decomposable like a tuple.